mlflow/utils/search_utils.py

```diff
@@ -303,7 +229,93 @@ def does_run_match_clause(cls, run, sed):
+        except Exception:
+            raise MlflowException("Error on parsing filter '%s'" % filter_string,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        if len(parsed) == 0 or not isinstance(parsed[0], Statement):
+            raise MlflowException("Invalid filter '%s'. Could not be parsed." %
+                                  filter_string, error_code=INVALID_PARAMETER_VALUE)
+        elif len(parsed) > 1:
+            raise MlflowException("Search filter contained multiple expression '%s'. "
+                                  "Provide AND-ed expression list." % filter_string,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        return SearchUtils._process_statement(parsed[0])
+
+    @classmethod
+    def filter(cls, runs, filter_string):
+        """Filters a set of runs based on a search filter string."""
+        if not filter_string:
+            return runs
+        parsed = cls._parse_search_filter(filter_string)
+
+        def run_matches(run):
+            return all([cls._does_run_match_clause(run, s) for s in parsed])
+        return [run for run in runs if run_matches(run)]
+
+    @classmethod
+    def _parse_order_by(cls, order_by):
+        try:
+            parsed = sqlparse.parse(order_by)
+        except Exception:
+            raise MlflowException("Error on parsing order_by clause '%s'" % order_by,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        if len(parsed) != 1 or not isinstance(parsed[0], Statement):
+            raise MlflowException("Invalid order_by clause '%s'. Could not be parsed." %
+                                  order_by, error_code=INVALID_PARAMETER_VALUE)
+
+        statement = parsed[0]
+        if len(statement.tokens) != 1 or not isinstance(statement[0], Identifier):
+            raise MlflowException("Invalid order_by clause '%s'. Could not be parsed." %
+                                  order_by, error_code=INVALID_PARAMETER_VALUE)
+
+        token_value = statement.tokens[0].value
+        is_ascending = True
+        if token_value.lower().endswith(" desc"):
+            is_ascending = False
+            token_value = token_value[0:-len(" desc")]
+        elif token_value.lower().endswith(" asc"):
+            token_value = token_value[0:-len(" false")]
+        identifier = cls._get_identifier(token_value)
+        return (identifier["type"], identifier["key"], is_ascending)
+
+    @classmethod
+    def _get_sort_key(cls, run, key_type, key, ascending):
+        sort_value = None
+        if key_type == cls._METRIC_IDENTIFIER:
+            sort_value = run.data.metrics.get(key)
+        elif key_type == cls._PARAM_IDENTIFIER:
+            sort_value = run.data.params.get(key)
+        elif key_type == cls._TAG_IDENTIFIER:
+            sort_value = run.data.tags.get(key)
+        elif key_type == cls._ATTRIBUTE_IDENTIFIER:
+            sort_value = getattr(run.info, key)
```

Comment:
catch error if invalid `key` used and throw appropriate `MlflowExpection` for usage of invalid key order_by

Reply:
Hmm, we don't do this in the search case, and my understanding was that that was because the key is already validated by the time we get here. Indeed, we do have tests that confirm `attribute.start` is disallowed.

Reply:
gotcha! thanks for that test case